### ISSUE TITLE
String binary support

### ISSUE THREAD

**Issue by kalinjul** [CreateAPI/CreateAPI#151]
Hi,
i'd like to use CreateApi, but my open api spec has the following item inside a multipart/form-data request:

`                "properties": {
                  "documents": {
                    "type": "array",
                    "items": {
                      "type": "string",
                      "format": "binary"
                    }
                  },
`

This generates to an Entity that has a String Array. But SwaggerUI sends plain bytes. I guess a String can't hold plain bytes, so I'd need to encode them somehow and this won't be accepted by the Backend. 

Is this a missing feature or am I missing something?

**Comment by liamnichols** [comment]
> inside a multipart/form-data request

I think this is likely to be the root issue here. CreateAPI is currently primarily focused on supporting JSON serialisation but I guess that you need to make a `multipart/form-data` request here? 

This wouldn't be supported in the simplest configuration (i.e using Get as an API Client) but there might be some ways that you can work around this although I haven't tried the below...

Firstly, try mapping the `binary` format to a different type in your configuration:

```yml
```

**Comment by liamnichols** [comment]
> inside a multipart/form-data request

I think this is likely to be the root issue here. CreateAPI is currently primarily focused on supporting JSON serialisation but I guess that you need to make a `multipart/form-data` request here? 

This wouldn't be supported in the simplest configuration (i.e using Get as an API Client) but there might be some ways that you can work around this although I haven't tried the below...

Firstly, try mapping the `binary` format to a different type in your configuration using [`dataTypes`](https://github.com/CreateAPI/CreateAPI/blob/main/Docs/ConfigOptions.md#datatypes):

```yml
generate: [paths, entities, enums]
dataTypes:
  string:
    binary: File
```

The generated code should then turn your `documents` array into `[File]`. You'll need to add your own type called `File` and it will need to conform to the `Encodable` and `Decodable` protocol to compile I think (you might be able to limit this by setting [`entities.alwaysIncludeDecodableImplementation`](https://github.com/CreateAPI/CreateAPI/blob/main/Docs/ConfigOptions.md#entitiesalwaysincludedecodableimplementation) to `false`).

Your `File` type could maybe be something like this:

```swift
public struct File: Codable {
    let fileURL: URL
    let contentType: String
}
```

While using `JSONEncoder` (or probably any other encoder) wouldn't encode this like you want, you could maybe write your own encoder, or somehow map the generated entities into something else that lets you create the HTTP Body that you need?

I'm not sure if there are any other third party libraries out there that help more, but let me know if any of this helps. If it doesn't, could you share a little more about how you plan to integrate CreateAPI (i.e to generate a Swift Package, using Get or a third party client etc)?

**Comment by kalinjul** [comment]
Thanks for your tips!
I'm trying to generate a Swift Package using Get for a non-public API. I have an OpenAPI spec.
The request is a multipart/form-data because it supports upload of multiple files (and has some other fields, too).

```
        "requestBody": {
          "content": {
            "multipart/form-data": {
                "type": "object",
                "properties": {
                  "documents": {
                    "type": "array",
                    "items": {
                      "type": "string",
                      "format": "binary"
                    },
                  },
                  "totalAmount": {
                    "type": "number",
                    "format": "double",
                    "example": 8.99
                  },
...
```

When I'm trying your solution, CreateApi doesn't change the [String] to [File] although it works on other fields. I guess it's not supported to change dataTypes inside Arrays?

**Comment by LePips** [comment]
Yes, that's correct that array overrides haven't been added. I implemented the data type overrides but those only apply to schema `dataType`s and not arrays because they are defined as `object`s. I'll admit I didn't take much look into arrays being overriden as well but I don't think it will be that difficult.

**Comment by kalinjul** [comment]
I stepped into the code and it looks like the array information is already lost while parsing.
I got into this line: https://github.com/CreateAPI/CreateAPI/blob/22074e5167c0b509a5241ba72a3cdbac894f88b1/Sources/CreateAPI/Generator/Generator%2BSchemas.swift#L593
and `info` is ArrayFormat.generic which means no type information.
I don't have time to dig any further, i suspect it's a missing feature in OpenAPIKit.

**Comment by LePips** [comment]
OpenAPIKit works as expected, the data type overrides are done by CreateAPI. We would implement this by looking at the type definition of the property being created and override that with the provided override from the config.

**Comment by liamnichols** [comment]
Its expected that `info.format` would be `.generic` given this schema: 

```json
{
  "documents" {
    "type": "array",
    "items": {
      "type": "string",
      "format": "binary"
    }
  }
}
```

Because the format we actually care about is the format of the nested items schema... At least that's what I would have thought 😅 

> We would implement this by looking at the type definition of the property being created and override that with the provided override from the config

I didn't run it, but that seems to be what should already happen. `makeTypealiasArray(...)` calls into `_makeDeclaration(...)` and passes the `items` schema which should already have the annotation. 

I need to run the code and play around a bit to see what is happening though

**Comment by liamnichols** [comment]
I just tried it and it worked as expected for me. Which version of CreateAPI are you using? `dataTypes` was only added to 0.1.0.

One thing to note is that when I used `multipart/form-data` for the content type, CreateAPI initially generated the request body as an entire `Data` for which I assume was because it doesn't technically support multipart form data encoding.

**Comment by kalinjul** [comment]
You're right, the multipart/form-data request is generated as Data, i actually overlooked this.
But i can't get the replacement of binary working for entities:
```yaml
  "components": {
    "schemas": {
      "Receipt": {
        "type": "object",
        "properties": {
          "documents": {
            "type": "array",
            "items": {
              "type": "string",
              "format": "binary"
            },
            "nullable": true
          }
        },
        "additionalProperties": false
      },
```

Together with
```yaml
dataTypes:
  string:
    binary: File
```

produces:
```yaml
public struct Receipt: Codable {
    public var documents: [String]?
```

Once i change the "format" from "binary" to "custom" and replace custom with binary in `dataTypes`, i get the correct result:
```yaml
public struct Receipt: Codable {
    public var documents: [File]?
```

But overall, what I'm missing is multipart support. Sorry for the misdirection. Is Get capable of multipart requests?

**Comment by liamnichols** [comment]
> But overall, what I'm missing is multipart support. Sorry for the misdirection

No troubles. This isn't something that I had considered much either but actually I am going to run into a similar issue on our project at some point in the future too 😄 

> Is Get capable of multipart requests?

I think there are workarounds, but I'm not sure if it's technically considered as supported. The `APIClient` in Get is designed to accept an `Encodable` body type which will then use a universal Encoder to encode into a HTTP body. 

You could theoretically work around this by using a different instance of `APIClient` that use a custom configuration with an alternative encoder, but currently it isn't very well joined together so it would still require some manual work. 

To bridge this information better the `Request` type might need a mechanism to define different types of encoding and decoding mechanisms that are supported. I'm not entirely sure how this would look though. 

Beyond Alamofire's multipart form data support, I haven't actually seen much support for multipart requests in Swift. If there was a good Encoder that builds from Swift's Codable system, I guess there could be potential to tell CreateAPI to generate a typed request body (like we do with `application/json`) and then somehow have a way to use different Encoders in Get's `APIClient` to turn that into the correct body data?

The next problem that I could potentially see us facing here however is dealing with large files. Since with large files, you might find that you need to use background upload tasks which Get certainly doesn't support.  

I'm not sure if there are other options either though? cc'ing @kean on this one too as he might have some thoughts.
